src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateClone = { ...state };
+  const stateHistory = [];
+
+  for (const obj of actions) {
+    if (obj.type === 'addProperties') {
+      Object.assign(stateClone, obj.extraData);
+      stateHistory.push({ ...stateClone });
+    }
+
+    if (obj.type === 'removeProperties') {
+      for (const item of obj.keysToRemove) {
+        delete stateClone[item];
+      }
+      stateHistory.push({ ...stateClone });
+    }
+
+    if (obj.type === 'clear') {
+      for (const key in stateClone) {
+        delete stateClone[key];
+      }
+      stateHistory.push({ ...stateClone });
```

Comment:
Consider moving this line outside of the `if` block to ensure that the state history is updated consistently after each action.